src/app/shared/modules/context-menu/context-menu.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable } from "@angular/core";
+import { PopoverController } from "@ionic/angular";
+import { fromEvent } from "rxjs";
+import { filter, first, map, switchMap, tap } from "rxjs/operators";
+import { ContextMenuComponent } from "./context-menu.component";
+import { BehaviorSubject } from "rxjs";
+import { Subscription } from "rxjs";
+import {
+  IContextMenuAction,
+  IContextMenuActionHashmap,
+  IContextMenuType,
+} from "./context-menu.types";
+import { Observable } from "rxjs";
+
+@Injectable({ providedIn: "root" })
+/**
+ * The context menu service tracks interaction with right-click and text-selection actions
+ * and presents a custom context menu where required
+ * TODO - add mobile long-press (and test on mobile). E.g. via getstureController, hammerjs or touch start/end events
+ * https://ionicframework.com/docs/utilities/gestures#overview
+ * https://levelup.gitconnected.com/making-hammerjs-work-with-angular-9-81d289159320
+ */
```

Comment:
This is the service that handles listening for events and triggering actions accordingly. The most interesting part is the definition of handlers (the rest is mostly just organisation of subscription services)